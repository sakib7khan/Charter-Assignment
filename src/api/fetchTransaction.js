export const fetchTransactions = async () => {
  const res = await fetch('http://localhost:3005/transactions');
  if (!res.ok) throw new Error('API fetch failed');
  return await res.json();
};
