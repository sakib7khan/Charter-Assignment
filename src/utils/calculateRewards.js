export const calculateRewards = (amount) => {
  let rewards = 0;

  if (amount > 100) {
    rewards += (amount - 100) * 2;
    rewards += 50; // for the $50â€“$100 range
  } else if (amount > 50) {
    rewards += (amount - 50) * 1;
  }

  return Math.floor(rewards);
};
