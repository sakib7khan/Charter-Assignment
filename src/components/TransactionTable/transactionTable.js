import React, { useEffect, useState } from 'react';
import { fetchTransactions } from '../../api/fetchTransaction';
import { calculateRewards } from '../../utils/calculateRewards';
import {
  Container,
  Title,
  Table,
  Button,
  Spinner,
  ErrorMessage,
  Info,
} from '../../styles/stylesComponents';

const TransactionTable = ({ customerId, selectedMonth, selectedYear }) => {
  const [transactions, setTransactions] = useState([]);
  const [page, setPage] = useState(1);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);
  const pageSize = 3;

  useEffect(() => {
    setLoading(true);
    setError(null);

    fetchTransactions()
      .then((data) => {
        let filtered = data.filter((txn) => txn.customerId === customerId);

        if (selectedMonth) {
          filtered = filtered.filter((txn) => {
            const txnMonth = new Date(txn.date).toLocaleString('default', {
              month: 'long',
            });
            return txnMonth === selectedMonth;
          });
        }

        if (selectedYear) {
          filtered = filtered.filter(
            (txn) =>
              new Date(txn.date).getFullYear().toString() === selectedYear
          );
        }

        setTransactions(filtered);
        setPage(1);
        setLoading(false);
      })
      .catch(() => {
        setError('Failed to load transactions.');
        setLoading(false);
      });
  }, [customerId, selectedMonth, selectedYear]);

  const paginatedTxns = transactions.slice(
    (page - 1) * pageSize,
    page * pageSize
  );

  const totalRewards = transactions.reduce(
    (acc, txn) => acc + calculateRewards(txn.amount),
    0
  );

  const customerName = transactions[0]?.customerName || customerId;

  return (
    <Container>
      <Title>Transactions for {customerName}</Title>
      {selectedMonth && (
        <Info>
          Filtered by Month: <strong>{selectedMonth}</strong>
        </Info>
      )}
      {selectedYear && (
        <Info>
          Filtered by Year: <strong>{selectedYear}</strong>
        </Info>
      )}

      {loading && <Spinner data-testid="loading-spinner">Loading transactions...</Spinner>}
      {error && <ErrorMessage>{error}</ErrorMessage>}

      {!loading && !error && transactions.length === 0 && (
        <Info>No transactions found for the selected filters.</Info>
      )}

      {!loading && !error && transactions.length > 0 && (
        <>
          <Info>Total Transactions: {transactions.length}</Info>
          <Info>Total Rewards: {totalRewards}</Info>

          <Table>
            <thead>
              <tr>
                <th>ID</th>
                <th>Amount ($)</th>
                <th>Date</th>
                <th>Rewards</th>
              </tr>
            </thead>
            <tbody>
              {paginatedTxns.map((txn) => (
                <tr key={txn.transactionId}>
                  <td>{txn.transactionId}</td>
                  <td>{txn.amount.toFixed(2)}</td>
                  <td>{txn.date}</td>
                  <td>{calculateRewards(txn.amount)}</td>
                </tr>
              ))}
            </tbody>
          </Table>

          <div style={{ marginTop: '1rem' }}>
            <Button
              data-testid='prev-button'
              onClick={() => setPage((p) => Math.max(p - 1, 1))}
              disabled={page === 1}
            >
              Prev
            </Button>
            <Button
              data-testid='next-button'
              onClick={() => setPage((p) => p + 1)}
              disabled={page * pageSize >= transactions.length}
            >
              Next
            </Button>
          </div>
        </>
      )}
    </Container>
  );
};

export default TransactionTable;
