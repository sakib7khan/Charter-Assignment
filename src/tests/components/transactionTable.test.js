import React from 'react';
import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import TransactionTable from '../../components/TransactionTable/transactionTable';
import { fetchTransactions } from '../../api/fetchTransaction';
import { calculateRewards } from '../../utils/calculateRewards';

jest.mock('../../api/fetchTransaction');
jest.mock('../../utils/calculateRewards');

const mockTransactions = [
  {
    transactionId: 'T1',
    customerId: 'C1',
    customerName: 'John Doe',
    amount: 120,
    date: '2023-08-10',
  },
  {
    transactionId: 'T2',
    customerId: 'C1',
    customerName: 'John Doe',
    amount: 80,
    date: '2023-08-15',
  },
  {
    transactionId: 'T3',
    customerId: 'C1',
    customerName: 'John Doe',
    amount: 150,
    date: '2023-08-20',
  },
  {
    transactionId: 'T4',
    customerId: 'C1',
    customerName: 'John Doe',
    amount: 200,
    date: '2023-08-25',
  },
];

describe('TransactionTable Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    calculateRewards.mockImplementation((amount) => Math.floor(amount / 10));
  });

  it('displays loading spinner initially', () => {
    fetchTransactions.mockResolvedValue([]);
    render(<TransactionTable customerId='C1' />);
    expect(screen.getByTestId('loading-spinner')).toBeInTheDocument();
  });

  it('renders transactions and calculates rewards correctly', async () => {
    fetchTransactions.mockResolvedValue(mockTransactions);
    render(<TransactionTable customerId='C1' />);

    await waitFor(() => {
      expect(screen.getByText('Transactions for John Doe')).toBeInTheDocument();
    });

    // 3 items on page 1 (pagination)
    expect(screen.getAllByRole('row')).toHaveLength(4); // including header

    // Check rewards calculated
    expect(screen.getByText('12')).toBeInTheDocument(); // 120 => 12
    expect(screen.getByText('8')).toBeInTheDocument(); // 80 => 8
    expect(screen.getByText('15')).toBeInTheDocument(); // 150 => 15

    // Pagination buttons
    expect(screen.getByTestId('prev-button')).toBeDisabled();
    expect(screen.getByTestId('next-button')).toBeEnabled();
  });

  it('handles pagination correctly', async () => {
    fetchTransactions.mockResolvedValue(mockTransactions);
    render(<TransactionTable customerId='C1' />);

    await waitFor(() => screen.getByText('Transactions for John Doe'));

    // Click next
    fireEvent.click(screen.getByTestId('next-button'));

    // Check new row on page 2
    await waitFor(() => {
      expect(screen.getByText('T4')).toBeInTheDocument();
    });

    expect(screen.getByTestId('prev-button')).toBeEnabled();
    expect(screen.getByTestId('next-button')).toBeDisabled();
  });

  it('matches the snapshot', () => {
    fetchTransactions.mockResolvedValue(mockTransactions);
    const { asFragment } = render(<TransactionTable customerId='C1' />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('filters by selected month and year', async () => {
    fetchTransactions.mockResolvedValue(mockTransactions);
    render(
      <TransactionTable
        customerId='C1'
        selectedMonth='August'
        selectedYear='2023'
      />
    );

    await waitFor(() => screen.getByText('Filtered by Month:'));
    expect(screen.getByText('Filtered by Year:')).toBeInTheDocument();

    await waitFor(() => {
      expect(screen.queryByTestId('loading-spinner')).not.toBeInTheDocument();
    });
    expect(screen.getAllByRole('row')).toHaveLength(4); // 3 txns on page 1
  });

  it('displays "no transactions" message when list is empty', async () => {
    fetchTransactions.mockResolvedValue([]);
    render(<TransactionTable customerId='C1' />);

    await waitFor(() => screen.getByText(/no transactions found/i));
  });

  it('handles fetch error', async () => {
    fetchTransactions.mockRejectedValue(new Error('API failed'));
    render(<TransactionTable customerId='C1' />);

    await waitFor(() => screen.getByText('Failed to load transactions.'));
  });
});
