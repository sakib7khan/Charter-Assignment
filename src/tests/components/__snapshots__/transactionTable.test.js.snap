// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionTable Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-Qotzb fbOokM"
  >
    <h1
      class="sc-fYsHOw gOtxjZ"
    >
      Transactions for C1
    </h1>
    <div
      class="sc-iQQCXo bezgzY"
      data-testid="loading-spinner"
    >
      Loading transactions...
    </div>
  </div>
</DocumentFragment>
`;
